### PR TITLE
Add more user-friendly CLI 

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation
Add a more user-friendly cli requested by issue https://github.com/InternLM/lmdeploy/issues/427

lists
- [x] Add a unified lmdeploy cli in `lmdeploy.cli`
- [x] replace cli examples in documents

## Modification

Add LMDeploy CLI
```
lmdeploy 
├── convert
├── lite
│   ├── calibrate
│   ├── kv_qparams
│   ├── auto_awq
├── chat
│   ├── torch
│   ├── turbomind
├── serve
│   ├── gradio # webui
│   ├── api_server # fastapi server
│   ├── api_client # fastapi client
│   ├── triton_client # triton client
```

```
NAME
    lmdeploy - LMDeploy Command Line Interface.

SYNOPSIS
    lmdeploy GROUP | COMMAND

DESCRIPTION
    The CLI provides a unified API for converting, compressing and deploying
    large language models.

GROUPS
    GROUP is one of the following:

     chat
       Chat through terminal with pytorch or turbomind model.

     lite
       CLI for compressing LLMs.

     serve
       Serve LLMs and interact on terminal or web UI.

COMMANDS
    COMMAND is one of the following:

     convert
       Convert LLMs to lmdeploy format.
```

## BC-breaking (Optional)

None
#### convert
~`python3 -m lmdeploy.serve.turbomind.deploy internlm-chat-7b /path/to/internlm-chat-7b`~
`lmdeploy convert /path/to/internlm-chat-7b`

#### Inference by turbomind
~`python3 -m lmdeploy.turbomind.chat ./workspace`~
`lmdeploy chat turbomind ./workspace`

#### Serve with gradio

~`python3 -m lmdeploy.serve.gradio.app ./workspace`~
`lmdeploy serve gradio ./workspace`

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by lvhan028** [comment]
I suggest we change `commands` to `tools`

File: docs/en/supported_models/codellama.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ Based on the above table, download the model that meets your requirements. Execu
 python3 -m pip install lmdeploy
 
 # convert weight layout
-python3 -m lmdeploy.serve.turbomind.deploy codellama /the/path/of/codellama/model
+lmdeploy serve convert codellama /the/path/of/codellama/model
```
Comment (AllentDan) [inline docs/en/supported_models/codellama.md:32]: serve convert?

Comment (RunningLeon) [reply docs/en/supported_models/codellama.md:32]: typo

File: README.md [hunk 2]
```diff
@@ -118,14 +118,14 @@ git clone https://huggingface.co/internlm/internlm-chat-7b-v1_1 /path/to/internl
...
-python3 -m lmdeploy.serve.turbomind.deploy internlm-chat-7b /path/to/internlm-chat-7b
+lmdeploy convert internlm-chat-7b /path/to/internlm-chat-7b
 
 ```
 
 #### Inference by TurboMind
 
 ```shell
-python -m lmdeploy.turbomind.chat ./workspace
+lmdeploy chat turbomind ./workspace
```
Comment (lvhan028) [inline README.md:128]: Can we change it to `lmdeploy turbomind ./workspace`?

**Comment by lvhan028** [comment]
@AllentDan 请测试 internlm-7b-v1.1，internlm-20b 的转换、推理和服务是否正确
@irexyc 请测试 qwen-7b, qwen-14b 的转换、推理和服务是否正确
@lvhan028 测试 puyu-123b, llama2 转换、量化、推理和服务是否正确

**Review by irexyc — APPROVED** [review]

**Review by AllentDan — APPROVED** [review]
LGTM

File: docs/en/w4a16.md [hunk 3]
```diff
@@ -26,22 +26,22 @@ As demonstrated in the command below, first convert the model's layout using `tu
...
+lmdeploy chat turbomind ./workspace
 ```
 
 ## Serve with gradio
 
 If you wish to interact with the model via web ui, please initiate the gradio server as indicated below:
 
 ```shell
-python3 -m lmdeploy.serve.turbomind ./workspace --server_name {ip_addr} ----server_port {port}
+lmdeploy serve gradio ./workspace --server_name {ip_addr} --server_port {port}
```
Comment (lvhan028) [inline docs/en/w4a16.md:44]: @AllentDan 执行这条命令可以成功，但是我在浏览器上聊天，会卡住。服务端会报一个错误：'AsyncRequest' object has no attribute '_json_response_data'。更新gradio可以成功。lmdeploy现在对gradio有最低版本要求么？
我原先用的版本是：gradio 3.39.0

File: docs/zh_cn/supported_models/codellama.md [hunk 4]
```diff
@@ -99,16 +99,16 @@ python3 -m lmdeploy.serve.openai.api_server ./workspace server_ip server_port --
...
 
 或者，启动 gradio，在 webui 的聊天对话框中，与 codellama 交流：
 
 ```shell
 # restful_api_url 就是 api_server 产生的，比如 http://localhost:23333
 # server_ip 和 server_port 是用来提供 gradio ui 访问服务的
-# 例子: python -m lmdeploy.serve.gradio.app http://localhost:23333 localhost 6006 --restful_api True
-python -m lmdeploy.serve.gradio.app restful_api_url server_ip --restful_api True
+# 例子: lmdeploy serve gradio http://localhost:23333 --server_name localhost --server_port 6006 --restful_api True
+lmdeploy serve gradio restful_api_url --server_name ${server_ip} --server_port ${port} --restful_api True
```
Comment (lvhan028) [inline docs/zh_cn/supported_models/codellama.md:111]: ${port} 与 93行的 ${server_port} 不一致

File: docs/en/supported_models/codellama.md [hunk 5]
```diff
@@ -90,23 +90,23 @@ Launch inference server by:
 ```shell
 # --instance_num: number of instances to performance inference, which can be viewed as max requests concurrency
 # --tp: the number of GPUs used in tensor parallelism
-python3 -m lmdeploy.serve.openai.api_server ./workspace server_ip server_port --instance_num 32 --tp 1
+lmdeploy serve api_server ./workspace --server_name ${server_ip} --server_port 6006 --instance_num 32 --tp 1
```
Comment (lvhan028) [inline docs/en/supported_models/codellama.md:93]: 这里固定6006不太合适，和中文文档保持一致吧

File: docs/en/restful_api.md [hunk 6]
```diff
@@ -3,7 +3,7 @@
 ### Launch Service
 
 ```shell
-python3 -m lmdeploy.serve.openai.api_server ./workspace 0.0.0.0 server_port --instance_num 32 --tp 1
+lmdeploy serve api_server ./workspace --server_name 0.0.0.0 --server_port 6006 --instance_num 32 --tp 1
```
Comment (lvhan028) [inline docs/en/restful_api.md:6]: 建议不要使用固定的 server_port

File: lmdeploy/cli/chat.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                        max_session_len=max_session_len,
+                        log_file=log_file,
+                        debug=debug,
+                        adapter=adapter)
+
+    def turbomind(self,
+                  model_path,
+                  session_id: int = 1,
+                  cap: str = 'chat',
+                  sys_instruct: str = None,
```
Comment (lvhan028) [inline lmdeploy/cli/chat.py:68]: sys_instruct去掉吧。在PR #591 中已经移除了。

File: lmdeploy/cli/chat.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            stream_output (bool): indicator for streaming output or not
+            **kwarg (dict): other arguments for initializing model's chat
+                template
+        """
+        from lmdeploy.turbomind.chat import main as run_turbomind_model
+
+        run_turbomind_model(model_path,
+                            session_id=session_id,
+                            cap=cap,
+                            sys_instruct=sys_instruct,
```
Comment (lvhan028) [inline lmdeploy/cli/chat.py:91]: May merge main

File: README_zh-CN.md [hunk 9]
```diff
@@ -149,23 +149,23 @@ python3 -m lmdeploy.serve.gradio.app ./workspace
...
 
 也可以通过 WebUI 方式来对话：
 
 ```shell
 # restful_api_url is what printed in api_server.py, e.g. http://localhost:23333
 # server_ip and server_port here are for gradio ui
-# example: python -m lmdeploy.serve.gradio.app http://localhost:23333 localhost 6006 --restful_api True
-python -m lmdeploy.serve.gradio.app restful_api_url server_ip --restful_api True
+# example: lmdeploy serve gradio http://localhost:23333 --server_name localhost --server_port 6006 --restful_api True
+lmdeploy serve gradio restful_api_url --server_name ${server_ip} --server_port ${port} --restful_api True
```
Comment (lvhan028) [inline README_zh-CN.md:168]: ${server_port}

File: docs/en/w4a16.md (hunk 3 above)
Comment (AllentDan) [reply docs/en/w4a16.md:44]: https://github.com/gradio-app/gradio/issues/4061#issuecomment-1540164621
根据这个回答，是的，不过我们默认 requirements.txt 安装最新版本 gradio

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**